src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -106,41 +390,313 @@ public function request(string $method, string $url, array $options = []): Respo
-            if ($response instanceof MockResponse) {
-                $mockResponses[] = $response;
+        return "metadata_{$requestHash}_{$variantKey}";
+    }
+
+    /**
+     * Build a variant key for caching, given an array of normalized headers and the vary fields.
+     *
+     * The key is a pipe-separated string of "header=value" pairs, with the special case of "header=" for headers that are not present.
+     *
+     * @param array<string, string|string[]> $normalizedHeaders
+     * @param string[]                       $varyFields
+     */
+    private function buildVariantKey(array $normalizedHeaders, array $varyFields): string
+    {
+        $parts = [];
+        foreach ($varyFields as $field) {
+            $lower = strtolower($field);
+            if (!isset($normalizedHeaders[$lower])) {
+                $parts[] = $field.'=';
             } else {
-                $clientResponses[] = $response;
+                $joined = \is_array($normalizedHeaders[$lower])
+                    ? implode(',', $normalizedHeaders[$lower])
+                    : $normalizedHeaders[$lower];
+                $parts[] = $field.'='.rawurlencode($joined);
             }
         }
 
-        if (!$mockResponses) {
-            return $this->client->stream($clientResponses, $timeout);
+        return implode('|', $parts);
+    }
+
+    /**
+     * Parse the Cache-Control header and return an array of directive names as keys
+     * and their values as values, or true if the directive has no value.
+     *
+     * @param string[] $header the Cache-Control header as an array of strings
+     *
+     * @return array<string, string|true> the parsed Cache-Control directives
+     */
+    private function parseCacheControlHeader(array $header): array
+    {
+        $parsed = [];
+        foreach ($header as $line) {
+            foreach (explode(',', $line) as $directive) {
+                if (str_contains($directive, '=')) {
+                    [$name, $value] = explode('=', $directive, 2);
+                    $parsed[trim($name)] = trim($value);
+                } else {
+                    $parsed[trim($directive)] = true;
+                }
+            }
         }
 
-        if (!$clientResponses) {
-            return new ResponseStream(MockResponse::stream($mockResponses, $timeout));
+        return $parsed;
+    }
+
+    /**
+     * Evaluates the freshness of a cached response based on its headers and expiration time.
+     *
+     * This method determines the state of the cached response by analyzing the Cache-Control
+     * directives and the expiration timestamp. It returns one of the following states:
+     * - 'fresh': if the cached response is still valid or has no expiration.
+     * - 'must-revalidate': if the response must be revalidated before use.
+     * - 'stale-but-usable': if the response is stale but can be used in case of errors.
+     * - 'stale': if the cached response is no longer valid or usable.
```

Comment:
I suggest creating private constants for those states, which will avoid having magic strings in multiple places (and will allow documenting this as `@return self::STATE_*`)

Reply:
adressed in https://github.com/symfony/symfony/pull/59576/commits/05970b866ae30596e43ab75ca433bbe0bb86044a